File: internal/agent/claude/session.go
// Package claude 实现 AgentSession（multi-agent 修订版 §4.1）。
//
// 默认 transport = claude-sdk-bridge（官方 Agent SDK 常驻桥）；桥不可用时回退 print
// （factory.go）。本包只依赖 internal/agent 的中性接口，不把 bridge/HTTP 细节泄漏给业务。
package claude

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"sync"

	"pieqi/internal/agent"
	"pieqi/internal/agent/claude/bridge"

	"github.com/google/uuid"
	"go.uber.org/zap"
)

// session 基于 claude-sdk-bridge 的 AgentSession 实现。
//
// 事件流：后台 goroutine 消费桥的 SSE（OpenEventStream），把桥事件翻译成中性
// agent.Event 投递给 OnEvent 回调，并维护 turn_end 结果表（按 clientRef）。
// Prompt 带 clientRef 发一轮，然后等自己那轮 turn_end 到达（bridge cancel 会补
// 合成 turn_end，故取消的轮也能返回，不挂死）。
type session struct {
	client *bridge.Client
	id     string
	cwd    string
	logger *zap.Logger

	ctx    context.Context
	cancel context.CancelFunc

	mu           sync.Mutex
	cond         *sync.Cond
	onEvent      func(agent.Event)
	closed       bool
	fatal        error
	turnEnds     map[string]turnOutcome // clientRef -> 结果
	lastResumeID string
}

// turnOutcome 一轮的结果：err 非空表示该轮失败/被取消。
type turnOutcome struct {
	err  error
	info *agent.TurnInfo
}

// 编译期断言：session 实现 AgentSession。
var _ agent.AgentSession = (*session)(nil)

func newSession(client *bridge.Client, id, cwd string, logger *zap.Logger) *session {
	if logger == nil {
		logger = zap.NewNop()
	}
	s := &session{
		client:   client,
		id:       id,
		cwd:      cwd,
		logger:   logger,
		turnEnds: make(map[string]turnOutcome),
	}
	s.ctx, s.cancel = context.WithCancel(context.Background())
	s.cond = sync.NewCond(&s.mu)
	return s
}

// ID 返回桥会话 id（会话生命周期内的稳定句柄；续问 id 经 turn_end 事件携带）。
func (s *session) ID() string { return s.id }

// ResumeID 返回最近一轮 turn_end 携带的 SDK resume id（续问句柄；未跑过轮时为空）。
// sessionBackedAdapter.RealSessionID 据此持久化真实可续问 id。
func (s *session) ResumeID() string {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.lastResumeID
}

// Caps 桥支持同进程多轮、可 resume、可流式。
func (s *session) Caps() agent.Caps {
	return agent.Caps{MultiTurnPersistent: true, ResumeSupported: true, Streaming: true}
}

// OnEvent 注册中性事件回调（最后一次为准）。
func (s *session) OnEvent(fn func(agent.Event)) {
	s.mu.Lock()
	s.onEvent = fn
	s.mu.Unlock()
}

// Prompt 发送一轮并阻塞到该轮结束（turn_end 按 clientRef 匹配）。
func (s *session) Prompt(ctx context.Context, text string) error {
	ref := uuid.New().String()
	if err := s.client.Prompt(ctx, s.id, text, ref); err != nil {
		return err
	}
	return s.waitTurn(ctx, ref)
}

// waitTurn 等 clientRef 对应轮的 turn_end（或致命错误/会话关闭/ctx 取消）。
func (s *session) waitTurn(ctx context.Context, ref string) error {
	// ctx 取消时唤醒 cond.Wait
	stop := make(chan struct{})
	go func() {
		select {
		case <-ctx.Done():
			s.cond.Broadcast()
		case <-stop:
		}
	}()
	defer close(stop)

	s.mu.Lock()
	defer s.mu.Unlock()
	for {
		if o, ok := s.turnEnds[ref]; ok {
			return o.err
		}
		if s.fatal != nil {
			return s.fatal
		}
		if s.closed {
			return errors.New("claude: session closed")
		}
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
		}
		s.cond.Wait()
	}
}

// Cancel 取消当前轮（interrupt；桥会为被中断的轮补合成 turn_end）。
func (s *session) Cancel(ctx context.Context) error {
	return s.client.Cancel(ctx, s.id)
}

// RespondPermission 审批响应（allow=true 批准）。
func (s *session) RespondPermission(ctx context.Context, reqID string, allow bool, optionID string) error {
	return s.client.RespondPermission(ctx, s.id, reqID, allow, optionID)
}

// Close 关闭会话：删桥会话（杀子进程）+ 终止事件循环。幂等。
func (s *session) Close(ctx context.Context) error {
	s.mu.Lock()
	already := s.closed
	s.closed = true
	s.cond.Broadcast()
	s.mu.Unlock()
	if already {
		return nil
	}
	s.cancel() // 终止 SSE 读取
	err := s.client.CloseSession(ctx, s.id)
	if err != nil {
		s.logger.Warn("claude: close bridge session", zap.Error(err))
	}
	return nil
}

// fail 记录致命错误（子进程崩溃/事件流断开）并唤醒等待者。
// 同时以 EventError 通知订阅者（sessionBackedAdapter 据此关闭 Done()，识别会话已死）。
func (s *session) fail(err error) {
	s.mu.Lock()
	if s.fatal == nil {
		s.fatal = err
	}
	s.cond.Broadcast()
	fn := s.onEvent
	s.mu.Unlock()
	if fn != nil {
		fn(agent.Event{Kind: agent.EventError, SessionID: s.id, Err: err})
	}
}

func (s *session) isClosed() bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.closed
}

// runEventLoop 消费桥 SSE 事件流直到结束。会话创建后由 factory 启动。
func (s *session) runEventLoop() {
	body, err := s.client.OpenEventStream(s.ctx, s.id)
	if err != nil {
		s.fail(fmt.Errorf("claude: open event stream: %w", err))
		return
	}
	defer body.Close()
	bridge.ConsumeSSE(s.ctx, body, func(kind string, raw []byte) {
		s.handleEvent(kind, raw)
	})
	// 流正常走到头：会话被桥关闭或桥重启。
	if !s.isClosed() {
		s.fail(bridge.ErrStreamClosed)
	}
}

// handleEvent 处理一条桥事件：翻译为中性 agent.Event 并投递；维护 turn_end 表。
func (s *session) handleEvent(kind string, raw []byte) {
	var ev bridge.SSEEvent
	if err := json.Unmarshal(raw, &ev); err != nil {
		s.logger.Warn("claude: bad sse frame", zap.String("kind", kind), zap.Error(err))
		return
	}
	ge := s.toAgentEvent(ev)

	s.mu.Lock()
	switch ev.Kind {
	case "turn_end":
		out := turnOutcome{info: toTurnInfo(ev.Turn)}
		if ev.IsError || ev.Subtype != "" && ev.Subtype != "success" {
			msg := ev.Message
			if msg == "" {
				msg = fmt.Sprintf("turn %s", ev.Subtype)
			}
			out.err = fmt.Errorf("claude: %s", msg)
		}
		if ev.ClientRef != "" {
			s.turnEnds[ev.ClientRef] = out
		}
		if ev.Turn != nil && ev.Turn.ResumeID != "" {
			s.lastResumeID = ev.Turn.ResumeID
		}
		s.cond.Broadcast()
	case "error":
		if s.fatal == nil {
			s.fatal = fmt.Errorf("claude: %s", ev.Message)
		}
		s.cond.Broadcast()
	}
	fn := s.onEvent
	s.mu.Unlock()

	if fn != nil {
		fn(ge)
	}
}

// toTurnInfo 把桥的 turn 载荷转成中性 TurnInfo。
func toTurnInfo(t *bridge.SSETurn) *agent.TurnInfo {
	if t == nil {
		return nil
	}
	return &agent.TurnInfo{ResumeID: t.ResumeID, Usage: t.Usage, CostUSD: t.CostUSD}
}

// toAgentEvent 把桥事件翻译成中性 agent.Event。
func (s *session) toAgentEvent(ev bridge.SSEEvent) agent.Event {
	e := agent.Event{
		Kind:      agent.EventKind(ev.Kind),
		TurnSeq:   ev.TurnSeq,
		SessionID: s.id,
	}
	switch ev.Kind {
	case "text_delta":
		e.Text = ev.Text
	case "thinking_delta":
		e.Text = ev.Text
		e.IsThought = true
	case "tool_start":
		e.ToolCallID = ev.ToolCallID
		e.ToolTitle = ev.ToolTitle
		e.ToolKind = ev.ToolKind
		e.RawInput = ev.RawInput
	case "tool_end":
		e.ToolCallID = ev.ToolCallID
		e.ToolTitle = ev.ToolTitle
		e.ToolStatus = ev.ToolStatus
		e.RawOutput = ev.RawOutput
	case "permission_needed":
		e.Permission = agent.PermissionRequest{
			ReqID:      ev.ReqID,
			SessionID:  s.id,
			ToolCallID: ev.ToolUseID,
			ToolTitle:  ev.ToolName,
			RawInput:   ev.RawInput,
		}
	case "turn_end":
		e.Turn = toTurnInfo(ev.Turn)
		if ev.IsError || (ev.Subtype != "" && ev.Subtype != "success") {
			e.Err = fmt.Errorf("claude: turn %s", ev.Subtype)
		}
	case "error":
		e.Err = fmt.Errorf("claude: %s", ev.Message)
	case "state_changed":
		e.State = ev.State
	}
	return e
}

File: services/claude-sdk-bridge/src/session.js
// src/session.js
// 一个 bridge 会话的运行时：把官方 Agent SDK 的常驻多轮 query 包装成
// 中性事件流 + 审批挂起/释放，供 HTTP 层（index.js）驱动。
//
// 关键设计（来自 spike 结论，见 docs/multi-agent-evaluation.md §8.4）：
//   - includePartialMessages: true —— 必需，否则没有 text_delta 增量，只有整块 assistant 消息
//   - settings.permissions.ask 强制走 ask 路径 —— 否则本机默认会直接放行 shell 工具
//   - canUseTool 返回挂起 Promise = 审批挂起；rid 由 host 经 respondPermission 释放
//   - signal abort 时释放挂起（cancel 不泄漏到下一轮）
//   - result 消息 = 一轮结束（turn_end），带 sdkSessionId / usage / cost
//   - 子进程崩溃 = query 抛错 → error 事件（不挂死）
import { query } from "@anthropic-ai/claude-agent-sdk";
import { randomUUID } from "node:crypto";

// 强制所有工具经 canUseTool 审批（host 侧决定放行/上卡）。
// 注：若该通配写法不被 SDK 接受（ask 规则校验失败），改回显式清单，如
// ['Bash', 'PowerShell', 'Edit', 'Write', 'WebFetch', 'Task', 'Agent']。
const ASK_ALL = ["*"];

// 只读/派生类工具自动放行：命中即在 _canUseTool 里直接 allow，不产生 permission_needed、
// 不占审批卡。（P5 修复）agent 一次并行多个只读工具（Grep/Glob/Read…）若都卡审批，
// 会叠加后端单决策槽互相覆盖 → 部分审批在 UI 不可见、任务假死到超时。只有变更类工具需要
// 人类审批，对齐 config.yaml 的 pieqi.hook_tools 意图（Bash/Write/Edit/NotebookEdit）。
const AUTO_ALLOW_TOOLS = new Set([
  "Grep", "Glob", "Read", "WebSearch", "WebFetch",
  "TodoWrite", "Agent", "Task", "Skill", "KillShell",
]);

export class SessionRuntime {
  /**
   * @param {object} opts
   * @param {string} opts.id            桥会话 id（uuid）
   * @param {string} opts.cwd           工作目录
   * @param {string|null} opts.resumeSdkSessionId 续问的 SDK 会话 id（可空）
   * @param {(ev: object) => void} opts.eventSink 事件出口（HTTP 层做广播 + 历史）
   * @param {object} [opts.logger]
   * @param {() => void} [opts.onClosed] 会话关闭回调（HTTP 层清理）
   */
  constructor(opts) {
    this.id = opts.id;
    this.cwd = opts.cwd;
    this.eventSink = opts.eventSink;
    this.logger = opts.logger ?? console;
    this.onClosed = opts.onClosed;

    this.state = "idle"; // idle | running | waiting_permission | closed
    this.sdkSessionId = opts.resumeSdkSessionId ?? null;
    this.turnSeq = 1; // 当前轮的序号；result 后 +1
    this.closed = false;

    this.queue = []; // 已入队的用户消息 {text, ref}（等待 SDK 拉取下一轮）
    this.waiters = []; // SDK 正等在 generator.next() 上的解析器
    this.pendingPerms = new Map(); // rid -> resolve(allow|deny)
    this.pendingTools = new Map(); // toolUseId -> {title, input}
    this.query = null;
    this.currentClientRef = null; // 当前正在处理的消息的 clientRef（turn_end 回带，host 精确关联）
    this.turnInFlight = false; // 有未结束的轮（cancel 据此补合成 turn_end）

    this._start();
  }

  // ---- 事件出口 ----

  _emit(kind, data = {}) {
    this.lastActivity = Date.now();
    this.eventSink({ kind, sessionId: this.id, turnSeq: this.turnSeq, state: this.state, ...data });
  }

  _setState(s) {
    if (this.state === s) return;
    this.state = s;
    this._emit("state_changed", { state: s });
  }

  // ---- SDK 启动与消费 ----

  _start() {
    const options = {
      cwd: this.cwd,
      permissionMode: "default",
      tools: { type: "preset", preset: "claude_code" },
      // 清掉继承的 allow 规则并强制所有工具走 ask → canUseTool（host 全权审批）
      settings: { permissions: { allow: [], deny: [], ask: ASK_ALL } },
      includePartialMessages: true,
      maxTurns: 200,
      canUseTool: (toolName, input, ctl) => this._canUseTool(toolName, input, ctl),
    };
    if (this.sdkSessionId) options.resume = this.sdkSessionId;
    this.query = query({ prompt: this.promptStream(), options });
    this._consume();
    this.logger.debug(`[bridge] session ${this.id} started (resume=${this.sdkSessionId ?? "none"}) cwd=${this.cwd}`);
  }

  async _consume() {
    try {
      for await (const msg of this.query) {
        this._onMessage(msg);
      }
      // 正常收尾（input 流结束）
      if (!this.closed) {
        this._emit("state_changed", { state: "closed" });
      }
    } catch (err) {
      // 子进程崩溃 / 连接断开：抛错而非挂死（spike U3 结论）
      if (!this.closed) {
        this.logger.error(`[bridge] session ${this.id} stream error: ${err?.message ?? err}`);
        this._emit("error", { message: String(err?.message ?? err) });
      }
    } finally {
      this._cleanup();
    }
  }

  _onMessage(msg) {
    switch (msg.type) {
      case "system": {
        // init 带 session 元数据（含 sdk session_id / model 等）；其他子类型忽略
        if (msg.session_id) this.sdkSessionId = msg.session_id;
        break;
      }
      case "stream_event": {
        const e = msg.event;
        if (e.type === "content_block_delta") {
          const d = e.delta;
          if (d?.type === "text_delta" && d.text) {
            this._emit("text_delta", { text: d.text, isThought: false });
          } else if (d?.type === "thinking_delta" && d.thinking) {
            this._emit("thinking_delta", { text: d.thinking, isThought: true });
          }
        }
        break;
      }
      case "assistant": {
        for (const block of msg.message?.content ?? []) {
          if (block.type === "tool_use") {
            this.pendingTools.set(block.id, { title: block.name, input: block.input ?? null });
            this._emit("tool_start", {
              toolCallId: block.id,
              toolTitle: block.name,
              toolKind: "",
              rawInput: block.input ?? null,
            });
          }
        }
        break;
      }
      case "user": {
        // tool_result 回显 → 对应该工具调用收尾
        const content = Array.isArray(msg.message?.content) ? msg.message.content : [];
        for (const block of content) {
          if (block.type === "tool_result" && block.tool_use_id) {
            const info = this.pendingTools.get(block.tool_use_id) ?? {};
            this.pendingTools.delete(block.tool_use_id);
            this._emit("tool_end", {
              toolCallId: block.tool_use_id,
              toolTitle: info.title ?? "",
              toolStatus: block.is_error === true ? "failed" : "completed",
              rawOutput: block.content ?? null,
            });
          }
        }
        break;
      }
      case "result": {
        // 一轮结束 = turn 边界（spike U1：每轮恰好一条 result）
        const cost = typeof msg.total_cost_usd === "number" ? msg.total_cost_usd : 0;
        this.sdkSessionId = msg.session_id ?? this.sdkSessionId;
        this.turnInFlight = false;
        this._emit("turn_end", {
          subtype: msg.subtype,
          isError: msg.is_error === true,
          clientRef: this.currentClientRef,
          turn: { resumeId: this.sdkSessionId, usage: msg.usage ?? null, costUsd: cost },
        });
        this.turnSeq += 1;
        this._setState("idle");
        break;
      }
      default:
        break;
    }
  }

  // ---- 多轮输入（streaming input 保活）----

  async *promptStream() {
    while (!this.closed) {
      const item = await this._nextPrompt();
      if (this.closed || item == null) return;
      this.currentClientRef = item.ref ?? null;
      this.turnInFlight = true;
      yield { type: "user", message: { role: "user", content: item.text }, parent_tool_use_id: null };
    }
  }

  _nextPrompt() {
    if (this.queue.length > 0) return Promise.resolve(this.queue.shift());
    return new Promise((resolve) => this.waiters.push(resolve));
  }

  /** 入队一轮用户消息（HTTP /prompt 调用）。clientRef 供 turn_end 回带，host 精确匹配本轮。关闭后返回 false。 */
  pushPrompt(text, clientRef) {
    if (this.closed) return false;
    const item = { text, ref: clientRef ?? null };
    if (this.waiters.length > 0) {
      this.waiters.shift()(item);
    } else {
      this.queue.push(item);
    }
    this._setState("running");
    return true;
  }

  // ---- 审批挂起/释放 ----

  _canUseTool(toolName, input, { toolUseID, requestId, signal }) {
    // 只读/派生类工具自动放行（不占审批卡；见 AUTO_ALLOW_TOOLS 注释）。
    if (AUTO_ALLOW_TOOLS.has(toolName)) {
      return Promise.resolve({ behavior: "allow", updatedInput: undefined });
    }
    const rid = requestId ?? toolUseID ?? randomUUID();
    return new Promise((resolve) => {
      this.pendingPerms.set(rid, resolve);
      if (signal) {
        // cancel / interrupt 时释放挂起（不泄漏到下一轮）
        signal.addEventListener(
          "abort",
          () => {
            if (this.pendingPerms.delete(rid)) {
              resolve({ behavior: "deny", message: "cancelled by host" });
            }
          },
          { once: true }
        );
      }
      this._setState("waiting_permission");
      this._emit("permission_needed", {
        reqId: rid,
        toolName,
        toolUseID,
        requestId,
        rawInput: input ?? null,
      });
    });
  }

  /** 审批响应（HTTP /permissions/:rid）。未找到 rid 返回 false。 */
  respondPermission(rid, allow, optionID) {
    const resolve = this.pendingPerms.get(rid);
    if (!resolve) return false;
    this.pendingPerms.delete(rid);
    if (allow) {
      resolve({ behavior: "allow", updatedInput: undefined });
    } else {
      resolve({ behavior: "deny", message: optionID ? `denied: ${optionID}` : "denied by user" });
    }
    this._setState("running");
    return true;
  }

  // ---- 取消 / 关闭 ----

  /** 取消当前轮（streaming-input 模式的 interrupt）。 */
  async cancel() {
    if (!this.query || this.closed) return;
    try {
      await this.query.interrupt();
    } catch (err) {
      this.logger.warn(`[bridge] session ${this.id} interrupt: ${err?.message ?? err}`);
    }
    // 被中断的轮若还没出 result，补一条合成 turn_end（host 的 Prompt 靠它返回，不挂死）
    if (this.turnInFlight) {
      this.turnInFlight = false;
      this._emit("turn_end", {
        subtype: "cancelled",
        isError: true,
        clientRef: this.currentClientRef,
        turn: { resumeId: this.sdkSessionId, usage: null, costUsd: 0 },
      });
      this.turnSeq += 1;
      this._setState("idle");
    }
  }

  /** 关闭会话：结束 input 流 + close() 杀子进程 + 释放挂起。幂等。 */
  close() {
    if (this.closed) return;
    this.closed = true;
    this._setState("closed");
    // 释放所有挂起审批（不让 canUseTool 死等）
    for (const resolve of this.pendingPerms.values()) {
      resolve({ behavior: "deny", message: "session closed" });
    }
    this.pendingPerms.clear();
    try {
      this.query?.close();
    } catch (err) {
      this.logger.warn(`[bridge] session ${this.id} close: ${err?.message ?? err}`);
    }
  }

  _cleanup() {
    this._setState("closed");
    this.onClosed?.();
  }
}
